lib/time/time.go

```diff
@@ -0,0 +1,493 @@
+	if err := starlark.UnpackPositionalArgs("from_timestamp", args, kwargs, 1, &x); err != nil {
+		return nil, err
+	}
+	return Time(time.Unix(x, 0)), nil
+}
+
+func now(thread *starlark.Thread, _ *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+	return Time(NowFunc()), nil
+}
+
+// Duration is a Starlark representation of a duration.
+type Duration time.Duration
+
+// assert at compile time that Duration implements Unpacker.
+var _ starlark.Unpacker = (*Duration)(nil)
+
+// Unpack is a custom argument unpacker
+func (d *Duration) Unpack(v starlark.Value) error {
+	switch x := v.(type) {
+	case Duration:
+		*d = x
+		return nil
+	case starlark.String:
+		dur, err := time.ParseDuration(string(x))
+		if err != nil {
+			return err
+		}
+
+		*d = Duration(dur)
+		return nil
+	}
+
+	return fmt.Errorf("got %s, want a duration, string, or int", v.Type())
+}
+
+// String implements the Stringer interface.
+func (d Duration) String() string { return time.Duration(d).String() }
+
+// Type returns a short string describing the value's type.
+func (d Duration) Type() string { return "time.duration" }
+
+// Freeze renders Duration immutable. required by starlark.Value interface
+// because duration is already immutable this is a no-op.
+func (d Duration) Freeze() {}
+
+// Hash returns a function of x such that Equals(x, y) => Hash(x) == Hash(y)
+// required by starlark.Value interface.
+func (d Duration) Hash() (uint32, error) {
+	return uint32(d) ^ uint32(int64(d)>>32), nil
+}
+
+// Truth reports whether the duration is non-zero.
+func (d Duration) Truth() starlark.Bool { return d != 0 }
+
+// Attr gets a value for a string attribute, implementing dot expression support
+// in starklark. required by starlark.HasAttrs interface.
+func (d Duration) Attr(name string) (starlark.Value, error) {
+	switch name {
+	case "hours":
+		return starlark.Float(time.Duration(d).Hours()), nil
+	case "minutes":
+		return starlark.Float(time.Duration(d).Minutes()), nil
+	case "seconds":
+		return starlark.Float(time.Duration(d).Seconds()), nil
+	case "milliseconds":
+		return starlark.MakeInt64(time.Duration(d).Milliseconds()), nil
+	case "microseconds":
+		return starlark.MakeInt64(time.Duration(d).Microseconds()), nil
+	case "nanoseconds":
+		return starlark.MakeInt64(time.Duration(d).Nanoseconds()), nil
+	}
+	return nil, fmt.Errorf("unrecognized %s attribute %q", d.Type(), name)
+}
+
+// AttrNames lists available dot expression strings. required by
+// starlark.HasAttrs interface.
+func (d Duration) AttrNames() []string {
+	return []string{
+		"hours",
+		"minutes",
+		"seconds",
+		"milliseconds",
+		"microseconds",
+		"nanoseconds",
+	}
+}
+
+// CompareSameType implements comparison of two Duration values. required by
+// starlark.Comparable interface.
+func (d Duration) CompareSameType(op syntax.Token, v starlark.Value, depth int) (bool, error) {
+	x := time.Duration(d)
+	y := time.Duration(v.(Duration))
+	switch op {
```

Comment:
return threeway(op, x, y), nil

Reply:
@adonovan I'm not sure, I get what you expect here. You would to see the content of the switch block moved into a new private function?

Reply:
Yes, and that function already exists below: it's `threeway`.

Reply:
@adonovan I probably miss something obvious but the current function has only 2 parameters while you are trying to call a function with 3 parameters. At best the second parameter could be somehow `x-y` but we could get an overflow.

Reply:
Oops, you're right; I'm not really using my brain. What I mean was this, which is less concise that I had imagined:
```
	cmp := 0
	if x, y := d, v.(Duration); x < y {
		cmp = -1
	} else if x > y {
		cmp = 1
	}
	return threeway(op, cmp)
```